From 1ec773948cfcd8e0633cddc9b31dd7e356210a4f Mon Sep 17 00:00:00 2001
From: Pradeep Kumar Chitrapu <quic_pradeepc@quicinc.com>
Date: Wed, 27 Jul 2022 12:48:36 -0700
Subject: [PATCH 08/12] ath12k: fix transmitting packets during recovery

Currently, the point where FW crash is asserted and point where
FW recovers, if during this time any packets arrive from mac80211
layer, ath12k_dp_tx() returns -ESHUTDOWN which leads to warning
print
  ath12k c000000.wifi: failed to transmit frame -108

If there is any traffic running directly or in the backbone, this
leads to flooding of the above print.

Add logic to handle packets in ath11k_mac_op_tx() itself if driver
is expecting recovery (ATH11K_FLAG_CRASH_FLUSH flag is set).

Signed-off-by: Pradeep Kumar Chitrapu <quic_pradeepc@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dp_tx.c | 3 +++
 drivers/net/wireless/ath/ath12k/mac.c   | 5 +++++
 2 files changed, 8 insertions(+)

--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -6745,6 +6745,11 @@ static void ath12k_mac_op_tx(struct ieee
 		return;
 	}
 
+	if (unlikely(test_bit(ATH12K_FLAG_CRASH_FLUSH, &ar->ab->dev_flags))) {
+		ieee80211_free_txskb(ar->ah->hw, skb);
+		return;
+	}
+
 	memset(skb_cb, 0, sizeof(*skb_cb));
 	skb_cb->vif = vif;
 
--- a/drivers/net/wireless/ath/ath12k/dp_rx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_rx.c
@@ -648,6 +648,9 @@ static int ath12k_dp_reo_cmd_send(struct
 	struct hal_srng *cmd_ring;
 	int cmd_num;
 
+	if (test_bit(ATH12K_FLAG_CRASH_FLUSH, &ab->dev_flags))
+		return -ESHUTDOWN;
+
 	cmd_ring = &ab->hal.srng_list[dp->reo_cmd_ring.ring_id];
 	cmd_num = ath12k_hal_reo_cmd_send(ab, cmd_ring, type, cmd);
 
